django/contrib/messages/storage/cookie.py

```diff
@@ -122,39 +136,67 @@ def _store(self, messages, response, remove_oldest=True, *args, **kwargs):
         returned), and add the not_finished sentinel value to indicate as much.
         """
         unstored_messages = []
-        encoded_data = self._encode(messages)
+
+        serialized_messages = MessagePartSerializer().dumps(messages)
+        encoded_data = self._encode_parts(serialized_messages)
         if self.max_cookie_size:
             # data is going to be stored eventually by SimpleCookie, which
             # adds its own overhead, which we must account for.
             cookie = SimpleCookie()  # create outside the loop
 
-            def stored_length(val):
-                return len(cookie.value_encode(val)[1])
+            def is_too_large_for_cookie(data):
+                return bool(
+                    data and len(cookie.value_encode(data)[1]) > self.max_cookie_size
+                )
 
-            while encoded_data and stored_length(encoded_data) > self.max_cookie_size:
+            def compute_msg(some_serialized_msg):
+                return self._encode_parts(
+                    some_serialized_msg + [json.dumps(self.not_finished)],
```

Comment:
`[json.dumps(self.not_finished)]` is always the same, can we pre-calculate it (maybe even in the `__init__()`).